src/components/assignment-panel/assignment-submission-info/assignment-submission-info.tsx

```diff
@@ -0,0 +1,174 @@
+    // but has since resubmitted (stale grade).
+    // This is independent of graded/submitted/unsubmitted
+    resubmitted: IStudent[]
+}
+
+interface AssignmentSubmissionInfoProps {
+
+}
+
+const Circle = ({ size=8, color, style, ...props }: { size?: number, color: string } & React.HTMLProps<HTMLDivElement>) => {
+    return (
+        <div style={{
+            width: size,
+            height: size,
+            borderRadius: "50%",
+            background: color,
+            ...style
+        }} { ...props } />
+    )
+}
+
+const LegendItem = ({ label, value, color, style, ...props }: { label: React.ReactNode, value: React.ReactNode, color: string } & React.HTMLProps<HTMLDivElement>) => {
+    return (
+        <div style={{ display: "flex", alignItems: "center", ...style }} { ...props }>
+            <Circle color={ color } style={{ marginRight: 4 }} />
+            { label }&nbsp;
+            <span style={{ color: "var(--jp-ui-font-color2)" }}>{ value }</span>
+        </div>
+    )
+}
+
+const SubmissionLegend = ({ graded, submitted, unsubmitted, resubmitted }: SubmissionLegendProps) => {
+    const gradedTitle = useMemo<string>(() => {
+        const resubmittedOnyens = new Set(resubmitted.map((s) => s.onyen))
+        return graded.map((s) => `${ s.onyen }${ resubmittedOnyens.has(s.onyen) ? " (stale)" : "" }`).join(", ")
+    }, [graded, resubmitted])
+    
+    return (
+        <div>
+            <div style={{ display: "flex", flexWrap: "wrap", columnGap: 8, fontSize: 13, marginTop: 2, marginBottom: 4 }}>
+                <Tooltip title={ gradedTitle }>
+                    <LegendItem label="Graded" value={ `${ graded.length }${ resubmitted.length > 0 ? "*" : ""}` } color="var(--md-green-500)" />
+                </Tooltip>
+                <Tooltip title={ submitted.map((s) => s.onyen).join(", ") }>
+                    <LegendItem label="Submitted" value={ submitted.length } color="var(--md-yellow-500)" />
+                </Tooltip>
+                <Tooltip title={ unsubmitted.map((s) => s.onyen).join(", ") }>
+                    <LegendItem label="Unsubmitted" value={ unsubmitted.length } color="var(--md-red-500)" />
+                </Tooltip>
+            </div>
+            { resubmitted.length > 0 && (
+                <div style={{ fontSize: 12, fontStyle: "italic", color: "var(--jp-ui-font-color2)" }}>
+                    *{ resubmitted.length } submission{ resubmitted.length > 1 ? "s are" : " is" } graded but out of date
+                </div>
+            ) }
+        </div>
+    )
+}
+
+export const AssignmentSubmissionInfo = ({ }: AssignmentSubmissionInfoProps) => {
+    const { assignment, students } = useAssignment()!
+
+    const [gradingActive, setGradingActive] = useState<boolean>(false)
+
+    const [graded, submitted, unsubmitted, resubmitted, total] = useMemo(() => {
+        if (!assignment || !students) return [[], [], [], [], 0]
+        let graded: IStudent[] = []
```

Comment:
I'm kinda confused, why are these variables of the type `IStudent[]`? Is that an array of interfaces of a student or of students' assignments? I would assume we would want the latter.

Reply:
It's really an array of Students, which implement IStudent. Originally it was an array of tuples (IStudent, ISubmission) but the submission wasn't actually needed anywhere and added further confusion.

Reply:
This whole memoized block just categorizes the list of students for a specific assignment into students who have a submission graded (possibly have a newer submission though), have submitted but haven't received any grade, and those who haven't submitted.